srt-protocol/src/listener/mod.rs

```diff
@@ -0,0 +1,324 @@
+use std::{collections::HashMap, io::Error, net::SocketAddr, time::Duration, time::Instant};
+
+use crate::{
+    connection::Connection,
+    packet::*,
+    protocol::{
+        pending_connection::{listen::Listen, ConnectionResult},
+        time::Timer,
+    },
+    settings::ConnInitSettings,
+};
+
+pub use crate::protocol::pending_connection::{AccessControlRequest, AccessControlResponse};
+
+#[derive(Clone, Debug)]
+pub struct ListenerSettings {}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct SessionId(pub SocketAddr, pub SocketId);
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ListenerStatistics {
+    rx_errors: u64,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Input {
+    Packet(ReceivePacketResult),
+    AccessResponse(Option<(SessionId, AccessControlResponse)>),
+    Timer,
+
+    PacketSent(usize),
+
+    ConnectionClosed(SessionId),
+    AccessRequested(SessionId),
+    ConnectionRejected(SessionId),
+    ConnectionOpened(SessionId),
+    ConnectionDropped(SessionId),
+    PacketDelegated(SessionId),
+    StatisticsUpdated,
+
+    Failure(ActionError),
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Action<'a> {
+    SendPacket((Packet, SocketAddr)),
+    RequestAccess(SessionId, AccessControlRequest),
+    RejectConnection(SessionId, Option<(Packet, SocketAddr)>),
+    OpenConnection(SessionId, Box<(Option<(Packet, SocketAddr)>, Connection)>),
+    DelegatePacket(SessionId, (Packet, SocketAddr)),
+    DropConnection(SessionId),
+    UpdateStatistics(&'a ListenerStatistics),
+    WaitForInput,
+    Close,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum ActionError {
+    SendPacketFailed,
+    ReceivePacketFailed,
+    SendStatistics,
+    RequestAccessFailed(SessionId),
+    PendingConnectionMissing(SessionId),
+    ActiveConnectionMissing(SessionId),
+    OpenConnectionFailed(SessionId),
+    DelegatePacketFailed(SessionId),
+}
+
+#[derive(Debug)]
+pub struct MultiplexListener {
+    settings: ConnInitSettings,
+    sessions: HashMap<SessionId, Option<Listen>>,
```

Comment:
Would be good to have a comment here indicating a value of None means the connection is already established

Reply:
I was going to introduce an explicit enum to delineate the meaning instead. Leverage the type system to declare meaning/intent where possible.